src/schema/v2/infiniteDiscovery/discoverArtworks.ts

```diff
@@ -47,21 +49,84 @@ export const DiscoverArtworks: GraphQLFieldConfig<void, ResolverContext> = {
         },
       }),
     },
+    useRelatedArtworks: { type: GraphQLBoolean, defaultValue: false },
   }),
   resolve: async (
     _root,
     args,
-    { weaviateCreateObjectLoader, weaviateGraphqlLoader, artworksLoader }
+    {
+      weaviateCreateObjectLoader,
+      weaviateGraphqlLoader,
+      artworksLoader,
+      relatedArtworksLoader,
+      marketingCollectionLoader,
+      savedArtworksLoader,
+    }
   ) => {
     if (
+      !artworksLoader ||
       !weaviateGraphqlLoader ||
-      !weaviateCreateObjectLoader ||
-      !artworksLoader
+      !weaviateCreateObjectLoader
     ) {
       new Error("A loader is not available")
     }
 
-    const { userId, limit = 5, offset = 0, certainty = 0.5, sort } = args
+    const {
+      userId,
+      limit = 5,
+      offset = 0,
+      certainty = 0.5,
+      sort,
+      useRelatedArtworks,
+    } = args
+
+    if (useRelatedArtworks) {
+      if (!savedArtworksLoader) {
+        return new Error("You need to be signed in to perform this action")
+      }
+
+      const { body: savedArtworks } = await savedArtworksLoader({
+        size: 28,
+        sort: "-position",
+        user_id: userId,
+        private: true,
+      })
+
+      // Extract saved artwork IDs
+      const savedArtworkIds = savedArtworks.map((artwork) => artwork.id)
+
+      // Load curated artworks
+      const curatedArtworksCollection = await marketingCollectionLoader(
+        "curators-picks"
+      )
+
+      const curatedArtworkIds = curatedArtworksCollection.artwork_ids
+
+      // Select two random artworks from curated artworks
+      const randomCuratedArtworksIds = sampleSize(curatedArtworkIds, 2)
+
+      const curatedArtworks = await artworksLoader({
+        ids: randomCuratedArtworksIds,
+      })
+
+      // use curated if there are no saved artworks
+      const finalArtworkIds =
+        savedArtworkIds.length > 0 ? [...savedArtworkIds] : curatedArtworkIds
+
+      // Limit the number of artwork IDs to a maximum of 30
+      const queryArtworkIds = finalArtworkIds.slice(0, 30)
```

Comment:
For users with many saved artworks, would this be the 28 _most recent_ saved works? I assume that's what the `-position` sort above is doing. If so, sounds good, we want to avoid using old saves.

Reply:
I don't know how the ES API performs with 30 works... we generally use just 1 work on the artwork screen, so will be interesting to see how performance scales... for QA probably not relevant (unless it times out), only important if we want to push to prod.

Reply:
> For users with many saved artworks, would this be the 28 most recent saved works? I assume that's what the -position sort above is doing. If so, sounds good, we want to avoid using old saves.

correct, `-position` filters recent. 

> I don't know how the ES API performs with 30 works... we generally use just 1 work on the artwork screen, so will be interesting to see how performance scales... for QA probably not relevant (unless it times out), only important if we want to push to prod.

[Here are some metrics](https://app.datadoghq.com/apm/resource/gravity.grape/grape.endpoint_run/43fc22fc9b94101d?query=env%3Astaging%20operation_name%3Agrape.endpoint_run%20resource_name%3A%22Api%3A%3AV1%3A%3ARelatedEndpoint%20GET%20%2Frelated%2Fartworks%22%20service%3Agravity.grape&env=staging&fromUser=false&resourceMapPrefs=qson%3A%28data%3A%28threshold%3A%210.1%29%2Cversion%3A%210%29&summary=qson%3A%28data%3A%28visible%3A%21t%2Cchanges%3A%28%29%2Cerrors%3A%28selected%3Acount%29%2Chits%3A%28selected%3Acount%29%2Clatency%3A%28selected%3Alatency%2Cslot%3A%28agg%3A95%29%2Cdistribution%3A%28isLogScale%3A%21f%29%2CshowTraceOutliers%3A%21t%29%2Csublayer%3A%28slot%3A%28layers%3Aservice%29%2Cselected%3Apercentage%29%2ClagMetrics%3A%28selectedMetric%3A%21s%2CselectedGroupBy%3A%21s%29%29%2Cversion%3A%211%29&traces=qson%3A%28data%3A%28%29%2Cversion%3A%210%29&start=1731322261263&end=1731408661263&paused=false), sometimes it takes up to ~4s. Lowering the amount of the input artworks should improve this metric.

Reply:
Great, thanks for the information. As I suspected it does look too slow to ship as is, but that's not a problem for QA, unless it's timing out. If we wanted to use ES we would probably need to look into increasing its resources, as we would want to be able to send over a few artworks in each query (but perhaps we could get away with 10, for example, as we may want to rotate the artworks in the query between batches).